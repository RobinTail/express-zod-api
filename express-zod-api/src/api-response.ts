import { z } from "zod";
import { isSchema } from "./common-helpers";
import { contentTypes } from "./content-type";

export const defaultStatusCodes = {
  positive: 200,
  negative: 400,
} satisfies Record<string, number>;

export type ResponseVariant = keyof typeof defaultStatusCodes;
export const responseVariants = Object.keys(
  defaultStatusCodes,
) as ResponseVariant[];

/** @public this is the user facing configuration */
export class ApiResponse<S extends z.ZodType> {
  readonly #schema: S;
  readonly #statusCodes?: [number, ...number[]];
  readonly #mimeTypes?: [string, ...string[]] | null;

  public constructor(
    subject:
      | S // shorthand for { schema: S }, default status codes and MIME types applied
      | {
          schema: S;
          /** @default 200 for a positive and 400 for a negative response */
          statusCode?: number | [number, ...number[]];
          /**
           * @example null is for no content, such as 204 and 302
           * @default "application/json"
           * */
          mimeType?: string | [string, ...string[]] | null;
        },
  ) {
    const { schema, statusCode, mimeType } = isSchema(subject)
      ? { schema: subject }
      : subject;
    this.#schema = schema;
    this.#statusCodes =
      typeof statusCode === "number" ? [statusCode] : statusCode;
    this.#mimeTypes = typeof mimeType === "string" ? [mimeType] : mimeType;
  }

  public normalize(fbStatus: number) {
    return {
      schema: this.#schema,
      statusCodes: this.#statusCodes ?? [fbStatus],
      mimeTypes:
        this.#mimeTypes === undefined ? [contentTypes.json] : this.#mimeTypes,
    };
  }
}

export type NormalizedResponse = ReturnType<
  ApiResponse<z.ZodType>["normalize"]
>;
