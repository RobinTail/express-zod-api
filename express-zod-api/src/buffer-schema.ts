import { z } from "zod/v4";

export const bufferSchema = z.custom<Buffer>(
  (subject) => Buffer.isBuffer(subject),
  { error: "Expected Buffer" },
);
