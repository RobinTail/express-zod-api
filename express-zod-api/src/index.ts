import "@express-zod-api/zod-plugin"; // side effects here
export { createConfig } from "./config-type.ts";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory.ts";
export { getMessageFromError } from "./common-helpers.ts";
export { ensureHttpError } from "./result-helpers.ts";
export { BuiltinLogger } from "./builtin-logger.ts";
export { Middleware } from "./middleware.ts";
export {
  ResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler.ts";
export { DependsOnMethod } from "./depends-on-method.ts";
export { ServeStatic } from "./serve-static.ts";
export { createServer, attachRouting } from "./server.ts";
export { Documentation } from "./documentation.ts";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors.ts";
export { testEndpoint, testMiddleware } from "./testing.ts";
export { Integration } from "./integration.ts";
export { EventStreamFactory } from "./sse.ts";

export { ez } from "./proprietary-schemas.ts";

// Convenience types
export type { Depicter } from "./documentation-helpers.ts";
export type { Producer } from "./zts-helpers.ts";

// Interfaces exposed for augmentation
export type { LoggerOverrides } from "./logger-helpers.ts";
export type { TagOverrides } from "./common-helpers.ts";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
import type {} from "qs"; // fixes TS2742 for attachRouting
export type { Routing } from "./routing.ts";
export type { FlatObject } from "./common-helpers.ts";
export type { Method } from "./method.ts";
export type { IOSchema } from "./io-schema.ts";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type.ts";
export type { ApiResponse } from "./api-response.ts";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  HeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security.ts";
