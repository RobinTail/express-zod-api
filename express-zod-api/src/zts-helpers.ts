import type ts from "typescript";
import type { FlatObject } from "./common-helpers.ts";
import type { SchemaHandler } from "./schema-walker.ts";

export interface ZTSContext extends FlatObject {
  isResponse: boolean;
  makeAlias: (key: object, produce: () => ts.TypeNode) => ts.TypeNode;
}

export type Producer = SchemaHandler<ts.TypeNode, ZTSContext>;
