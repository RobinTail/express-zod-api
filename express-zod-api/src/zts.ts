import type {
  $ZodArray,
  $ZodCatch,
  $ZodDefault,
  $ZodDiscriminatedUnion,
  $ZodEnum,
  $ZodIntersection,
  $ZodLazy,
  $ZodLiteral,
  $ZodNullable,
  $ZodObject,
  $ZodOptional,
  $ZodPipe,
  $ZodReadonly,
  $ZodRecord,
  $ZodString,
  $ZodTransform,
  $ZodTuple,
  $ZodUnion,
} from "@zod/core";
import * as R from "ramda";
import ts from "typescript";
import { globalRegistry, z } from "zod";
import { doesAccept, getTransformedType, isSchema } from "./common-helpers";
import { ezDateInBrand } from "./date-in-schema";
import { ezDateOutBrand } from "./date-out-schema";
import { ezFileBrand, FileSchema } from "./file-schema";
import { ProprietaryBrand } from "./proprietary-schemas";
import { ezRawBrand, RawSchema } from "./raw-schema";
import { FirstPartyKind, HandlingRules, walkSchema } from "./schema-walker";
import {
  ensureTypeNode,
  isPrimitive,
  makeInterfaceProp,
  makeLiteralType,
} from "./typescript-api";
import { Producer, ZTSContext } from "./zts-helpers";

const { factory: f } = ts;

const samples = {
  [ts.SyntaxKind.AnyKeyword]: "",
  [ts.SyntaxKind.BigIntKeyword]: BigInt(0),
  [ts.SyntaxKind.BooleanKeyword]: false,
  [ts.SyntaxKind.NumberKeyword]: 0,
  [ts.SyntaxKind.ObjectKeyword]: {},
  [ts.SyntaxKind.StringKeyword]: "",
  [ts.SyntaxKind.UndefinedKeyword]: undefined,
} satisfies Partial<Record<ts.KeywordTypeSyntaxKind, unknown>>;

const nodePath = {
  name: R.path([
    "name" satisfies keyof ts.TypeElement,
    "text" satisfies keyof Exclude<
      NonNullable<ts.TypeElement["name"]>,
      ts.ComputedPropertyName
    >,
  ]),
  type: R.path(["type" satisfies keyof ts.PropertySignature]),
  optional: R.path(["questionToken" satisfies keyof ts.TypeElement]),
};

const onLiteral: Producer = ({ _zod: { def } }: $ZodLiteral) => {
  const values = def.values.map((entry) =>
    entry === undefined
      ? ensureTypeNode(ts.SyntaxKind.UndefinedKeyword)
      : makeLiteralType(entry),
  );
  return values.length === 1 ? values[0] : f.createUnionTypeNode(values);
};

/**
 * @todo determine of makeAlias needed
 * @see https://stackoverflow.com/questions/8591873/determine-if-a-javascript-property-has-a-getter-or-setter-defined
 */
const onObject: Producer = (obj: $ZodObject, { isResponse, next, makeAlias }) =>
  makeAlias(obj, () => {
    const members = Object.entries(obj._zod.def.shape).map<ts.TypeElement>(
      ([key, value]) => {
        const isOptional = isResponse
          ? isSchema<$ZodOptional>(value, "optional")
          : doesAccept(value, undefined);
        const { description: comment, deprecated: isDeprecated } =
          globalRegistry.get(value) || {};
        return makeInterfaceProp(key, next(value), {
          comment,
          isDeprecated,
          isOptional,
        });
      },
    );
    return f.createTypeLiteralNode(members);
  });

const onArray: Producer = ({ _zod: { def } }: $ZodArray, { next }) =>
  f.createArrayTypeNode(next(def.element));

const onEnum: Producer = ({ _zod: { def } }: $ZodEnum) =>
  f.createUnionTypeNode(Object.values(def.entries).map(makeLiteralType));

const onSomeUnion: Producer = (
  { _zod: { def } }: $ZodUnion | $ZodDiscriminatedUnion,
  { next },
) => {
  const nodes = new Map<ts.TypeNode | ts.KeywordTypeSyntaxKind, ts.TypeNode>();
  for (const option of def.options) {
    const node = next(option);
    nodes.set(isPrimitive(node) ? node.kind : node, node);
  }
  return f.createUnionTypeNode(Array.from(nodes.values()));
};

const makeSample = (produced: ts.TypeNode) =>
  samples?.[produced.kind as keyof typeof samples];

const onOptional: Producer = ({ _zod: { def } }: $ZodOptional, { next }) =>
  f.createUnionTypeNode([
    next(def.innerType),
    ensureTypeNode(ts.SyntaxKind.UndefinedKeyword),
  ]);

const onNullable: Producer = ({ _zod: { def } }: $ZodNullable, { next }) =>
  f.createUnionTypeNode([next(def.innerType), makeLiteralType(null)]);

const onTuple: Producer = ({ _zod: { def } }: $ZodTuple, { next }) =>
  f.createTupleTypeNode(
    def.items
      .map(next)
      .concat(def.rest === null ? [] : f.createRestTypeNode(next(def.rest))),
  );

const onRecord: Producer = ({ _zod: { def } }: $ZodRecord, { next }) =>
  ensureTypeNode("Record", [def.keyType, def.valueType].map(next));

const intersect = R.tryCatch(
  (nodes: ts.TypeNode[]) => {
    if (!nodes.every(ts.isTypeLiteralNode)) throw new Error("Not objects");
    const members = R.chain(R.prop("members"), nodes);
    const uniqs = R.uniqWith((...props) => {
      if (!R.eqBy(nodePath.name, ...props)) return false;
      if (R.both(R.eqBy(nodePath.type), R.eqBy(nodePath.optional))(...props))
        return true;
      throw new Error("Has conflicting prop");
    }, members);
    return f.createTypeLiteralNode(uniqs);
  },
  (_err, nodes) => f.createIntersectionTypeNode(nodes),
);

const onIntersection: Producer = (
  { _zod: { def } }: $ZodIntersection,
  { next },
) => intersect([def.left, def.right].map(next));

const onPrimitive =
  (syntaxKind: ts.KeywordTypeSyntaxKind): Producer =>
  () =>
    ensureTypeNode(syntaxKind);

const onWrapped: Producer = (
  { _zod: { def } }: $ZodReadonly | $ZodCatch | $ZodDefault,
  { next },
) => next(def.innerType);

const onPipeline: Producer = (
  { _zod: { def } }: $ZodPipe,
  { next, isResponse },
) => {
  const target = def[isResponse ? "out" : "in"];
  const opposite = def[isResponse ? "in" : "out"];
  if (!isSchema<$ZodTransform>(target, "transform")) return next(target);
  const opposingType = next(opposite);
  const targetType = getTransformedType(target, makeSample(opposingType));
  const resolutions: Partial<
    Record<NonNullable<typeof targetType>, ts.KeywordTypeSyntaxKind>
  > = {
    number: ts.SyntaxKind.NumberKeyword,
    bigint: ts.SyntaxKind.BigIntKeyword,
    boolean: ts.SyntaxKind.BooleanKeyword,
    string: ts.SyntaxKind.StringKeyword,
    undefined: ts.SyntaxKind.UndefinedKeyword,
    object: ts.SyntaxKind.ObjectKeyword,
  };
  return ensureTypeNode(
    (targetType && resolutions[targetType]) || ts.SyntaxKind.AnyKeyword,
  );
};

const onNull: Producer = () => makeLiteralType(null);

const onLazy: Producer = ({ _zod: { def } }: $ZodLazy, { makeAlias, next }) =>
  makeAlias(def.getter, () => next(def.getter()));

const onFile: Producer = (schema: FileSchema) => {
  const stringType = ensureTypeNode(ts.SyntaxKind.StringKeyword);
  const bufferType = ensureTypeNode("Buffer");
  const unionType = f.createUnionTypeNode([stringType, bufferType]);
  return isSchema<$ZodString>(schema, "string")
    ? stringType
    : isSchema<$ZodUnion>(schema, "union")
      ? unionType
      : bufferType;
};

const onRaw: Producer = (schema: RawSchema, { next }) =>
  next(schema._zod.def.shape.raw);

const producers: HandlingRules<
  ts.TypeNode,
  ZTSContext,
  FirstPartyKind | ProprietaryBrand
> = {
  string: onPrimitive(ts.SyntaxKind.StringKeyword),
  number: onPrimitive(ts.SyntaxKind.NumberKeyword),
  bigint: onPrimitive(ts.SyntaxKind.BigIntKeyword),
  boolean: onPrimitive(ts.SyntaxKind.BooleanKeyword),
  any: onPrimitive(ts.SyntaxKind.AnyKeyword),
  undefined: onPrimitive(ts.SyntaxKind.UndefinedKeyword),
  [ezDateInBrand]: onPrimitive(ts.SyntaxKind.StringKeyword),
  [ezDateOutBrand]: onPrimitive(ts.SyntaxKind.StringKeyword),
  null: onNull,
  array: onArray,
  tuple: onTuple,
  record: onRecord,
  object: onObject,
  literal: onLiteral,
  intersection: onIntersection,
  union: onSomeUnion,
  default: onWrapped,
  enum: onEnum,
  optional: onOptional,
  nullable: onNullable,
  catch: onWrapped,
  pipe: onPipeline,
  lazy: onLazy,
  readonly: onWrapped,
  [ezFileBrand]: onFile,
  [ezRawBrand]: onRaw,
};

export const zodToTs = (
  schema: z.ZodType,
  {
    brandHandling,
    ctx,
  }: {
    brandHandling?: HandlingRules<ts.TypeNode, ZTSContext>;
    ctx: ZTSContext;
  },
) =>
  walkSchema(schema, {
    rules: { ...brandHandling, ...producers },
    onMissing: () => ensureTypeNode(ts.SyntaxKind.AnyKeyword),
    ctx,
  });
