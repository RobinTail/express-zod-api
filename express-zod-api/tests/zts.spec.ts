import ts from "typescript";
import { z } from "zod";
import { ez } from "../src";
import { f, printNode } from "../src/typescript-api";
import { zodToTs } from "../src/zts";
import { ZTSContext } from "../src/zts-helpers";

describe("zod-to-ts", () => {
  const printNodeTest = (node: ts.Node) =>
    printNode(node, { newLine: ts.NewLineKind.LineFeed });
  const ctx: ZTSContext = {
    isResponse: false,
    makeAlias: vi.fn(() => f.createTypeReferenceNode("SomeType")),
  };

  describe("z.array()", () => {
    test("outputs correct typescript", () => {
      const node = zodToTs(
        z.object({ id: z.number(), value: z.string() }).array(),
        { ctx },
      );
      expect(printNodeTest(node)).toMatchSnapshot();
    });
  });

  describe.each(["string", "base64", "binary", "buffer"] as const)(
    "ez.file(%s)",
    (variant) => {
      test("should depend on variant", () => {
        const node = zodToTs(ez.file(variant), { ctx });
        expect(printNodeTest(node)).toMatchSnapshot();
      });
    },
  );

  describe("ez.raw()", () => {
    test("should depict the raw property", () => {
      const node = zodToTs(ez.raw({ extra: z.string() }), { ctx });
      expect(printNodeTest(node)).toMatchSnapshot();
    });
  });

  describe("enums", () => {
    // noinspection JSUnusedGlobalSymbols
    enum Color {
      Red,
      Green,
      Blue,
    }

    // noinspection JSUnusedGlobalSymbols
    enum Fruit {
      Apple = "apple",
      Banana = "banana",
      Cantaloupe = "cantaloupe",
    }

    // noinspection JSUnusedGlobalSymbols
    enum StringLiteral {
      "Two Words",
      "'Quotes\"",
      '\\"Escaped\\"',
    }

    test.each([
      { schema: z.enum(Color), feature: "numeric" },
      { schema: z.enum(Fruit), feature: "string" },
      { schema: z.enum(StringLiteral), feature: "quoted string" },
    ])("handles $feature literals", ({ schema }) => {
      expect(printNodeTest(zodToTs(schema, { ctx }))).toMatchSnapshot();
    });
  });

  describe("Example", () => {
    // noinspection JSUnusedGlobalSymbols
    enum Fruits {
      Apple = "apple",
      Banana = "banana",
      Cantaloupe = "cantaloupe",
      A = 5,
    }

    const pickedSchema = z
      .object({
        string: z.string(),
        number: z.number(),
        fixedArrayOfString: z.array(z.string()).nonempty().length(10),
        object: z.object({
          string: z.string(),
        }),
      })
      .partial();

    const circular: z.ZodLazy = z.lazy(() =>
      z.object({
        a: z.string(),
        b: circular,
      }),
    );

    const circular2 = z.interface({
      name: z.string(),
      get subcategories() {
        return z.array(circular2);
      },
    });

    const example = z.object({
      string: z.string(),
      number: z.number(),
      arrayOfObjects: z.array(
        z.object({
          string: z.string(),
        }),
      ),
      boolean: z.boolean(),
      circular,
      circular2,
      union: z.union([z.object({ number: z.number() }), z.literal("hi")]),
      enum: z.enum(["hi", "bye"]),
      intersectionWithTransform: z
        .number()
        .and(z.bigint())
        .and(z.number().and(z.string()))
        .transform((arg) => console.log(arg)),
      date: z.date(),
      undefined: z.undefined(),
      null: z.null(),
      void: z.void(),
      any: z.any(),
      unknown: z.unknown(),
      never: z.never(),
      optionalString: z.optional(z.string()),
      nullablePartialObject: z.nullable(pickedSchema),
      tuple: z.tuple([
        z.string(),
        z.number(),
        z.object({ string: z.string() }),
      ]),
      tupleRest: z.tuple([z.string(), z.number()]).rest(z.boolean()),
      record: z.record(
        z.string(),
        z.object({
          object: z.object({
            arrayOfUnions: z
              .union([
                z.tuple([z.string(), z.object({ string: z.string() })]),
                z.bigint(),
              ])
              .array(),
          }),
        }),
      ),
      map: z.map(z.string(), z.array(z.object({ string: z.string() }))),
      set: z.set(z.string()),
      intersection: z.string().and(z.number()).or(z.bigint()),
      promise: z.promise(z.number()),
      optDefaultString: z.string().optional().default("hi"),
      refinedStringWithSomeBullshit: z
        .string()
        .refine((val) => val.length > 10)
        .or(z.number())
        .and(z.bigint().nullish().default(1000n)),
      nativeEnum: z.enum(Fruits),
      lazy: z.lazy(() => z.string()),
      discUnion: z.discriminatedUnion("kind", [
        z.object({ kind: z.literal("circle"), radius: z.number() }),
        z.object({ kind: z.literal("square"), x: z.number() }),
        z.object({ kind: z.literal("triangle"), x: z.number(), y: z.number() }),
      ]),
      branded: z.string().brand("BRAND"),
      catch: z.number().catch(123),
      pipeline: z.string().regex(/\d+/).transform(Number).pipe(z.number()),
      readonly: z.string().readonly(),
    });

    test("should produce the expected results", () => {
      const node = zodToTs(example, { ctx });
      expect(printNode(node)).toMatchSnapshot();
    });
  });

  describe("z.optional()", () => {
    const optionalStringSchema = z.string().optional();
    const objectWithOptionals = z.object({
      optional: optionalStringSchema,
      required: z.string(),
      transform: z
        .number()
        .optional()
        .transform((arg) => arg),
      or: z.number().optional().or(z.string()),
      tuple: z
        .tuple([
          z.string().optional(),
          z.number(),
          z.object({
            optional: z.string().optional(),
            required: z.string(),
          }),
        ])
        .optional(),
    });

    test("outputs correct typescript", () => {
      const node = zodToTs(optionalStringSchema, { ctx });
      expect(printNodeTest(node)).toMatchSnapshot();
    });

    test("should output `?:` and undefined union for optional properties", () => {
      const node = zodToTs(objectWithOptionals, { ctx });
      expect(printNodeTest(node)).toMatchSnapshot();
    });
  });

  describe("z.nullable()", () => {
    const nullableUsernameSchema = z.object({
      username: z.string().nullable(),
    });
    const node = zodToTs(nullableUsernameSchema, { ctx });

    test("outputs correct typescript", () => {
      expect(printNodeTest(node)).toMatchSnapshot();
    });
  });

  describe("z.object()", () => {
    test("supports string literal properties", () => {
      const schema = z.object({
        "string-literal": z.string(),
        5: z.number(),
      });
      const node = zodToTs(schema, { ctx });
      expect(printNodeTest(node)).toMatchSnapshot();
    });

    test("does not unnecessary quote identifiers", () => {
      const schema = z.object({
        id: z.string(),
        name: z.string(),
        countryOfOrigin: z.string(),
      });
      const node = zodToTs(schema, { ctx });
      expect(printNodeTest(node)).toMatchSnapshot();
    });

    test("escapes correctly", () => {
      const schema = z.object({
        "\\": z.string(),
        '"': z.string(),
        "'": z.string(),
        "`": z.string(),
        "\n": z.number(),
        $e: z.any(),
        "4t": z.any(),
        _r: z.any(),
        "-r": z.undefined(),
      });
      const node = zodToTs(schema, { ctx });
      expect(printNodeTest(node)).toMatchSnapshot();
    });

    test("supports zod.describe()", () => {
      const schema = z.object({
        name: z.string().describe("The name of the item"),
        price: z.number().describe("The price of the item"),
      });
      const node = zodToTs(schema, { ctx });
      expect(printNodeTest(node)).toMatchSnapshot();
    });

    test("supports zod.deprecated()", () => {
      const schema = z.object({
        one: z.string().deprecated(),
        two: z.string().deprecated().describe("with description"),
      });
      const node = zodToTs(schema, { ctx });
      expect(printNodeTest(node)).toMatchSnapshot();
    });

    test("specially handles coercive schema in response", () => {
      const schema = z.object({
        prop: z.coerce.string(),
      });
      const node = zodToTs(schema, { ctx: { ...ctx, isResponse: true } });
      expect(printNodeTest(node)).toMatchSnapshot();
    });
  });

  describe("Issue #2352: intersection of objects having same prop %#", () => {
    test.each([
      [z.string(), z.string()],
      [z.string().nonempty(), z.email()],
      [
        z.string().transform(Number),
        z
          .string()
          .transform((str) => new Date(str))
          .pipe(z.date()),
      ],
      [z.object({}), z.object({})],
    ])("should deduplicate the prop with a same name", (a, b) => {
      const schema = z.object({ query: a }).and(z.object({ query: b }));
      const node = zodToTs(schema, { ctx });
      expect(printNodeTest(node)).toMatchSnapshot();
    });

    test.each([
      [
        z.object({ query: z.string() }),
        z.object({ query: z.number() }), // prop type
      ],
      [
        z.object({ query: z.object({ sub: z.string() }) }),
        z.object({ query: z.object({ sub: z.number() }) }), // different complex
      ],
      [
        z.object({ query: z.string() }),
        z.object({ query: z.string().optional() }), // question mark
      ],
      [
        z.object({ query: z.string() }),
        z.object({ query: z.string() }).partial(), // also question mark
      ],
    ])(
      "should not flatten the result for objects with a conflicting prop %#",
      (a, b) => {
        const node = zodToTs(a.and(b), { ctx });
        expect(printNodeTest(node)).toMatchSnapshot();
      },
    );
  });

  describe("PrimitiveSchema", () => {
    const primitiveSchema = z.object({
      string: z.string(),
      number: z.number(),
      boolean: z.boolean(),
      date: z.date(),
      undefined: z.undefined(),
      null: z.null(),
      void: z.void(),
      any: z.any(),
      unknown: z.unknown(),
      never: z.never(),
    });
    const node = zodToTs(primitiveSchema, { ctx });

    test("outputs correct typescript", () => {
      expect(printNodeTest(node)).toMatchSnapshot();
    });
  });

  describe("z.discriminatedUnion()", () => {
    const shapeSchema = z.discriminatedUnion("kind", [
      z.object({ kind: z.literal("circle"), radius: z.number() }),
      z.object({ kind: z.literal("square"), x: z.number() }),
      z.object({ kind: z.literal("triangle"), x: z.number(), y: z.number() }),
    ]);
    const node = zodToTs(shapeSchema, { ctx });

    test("outputs correct typescript", () => {
      expect(printNodeTest(node)).toMatchSnapshot();
    });
  });

  describe("z.literal()", () => {
    test.each([
      z.literal("test"),
      z.literal(true),
      z.literal(false),
      z.literal(123),
      z.literal(undefined),
    ])("Should produce the correct typescript %#", (schema) => {
      expect(printNodeTest(zodToTs(schema, { ctx }))).toMatchSnapshot();
    });
  });

  describe("z.effect()", () => {
    describe("transformations", () => {
      test.each([
        { isResponse: false, expected: "intact" },
        { isResponse: true, expected: "transformed" },
      ])("should produce the schema type $expected", ({ isResponse }) => {
        const schema = z.number().transform((num) => `${num}`);
        expect(
          printNodeTest(zodToTs(schema, { ctx: { ...ctx, isResponse } })),
        ).toMatchSnapshot();
      });

      test("should handle unsupported transformation in response", () => {
        const schema = z.number().transform((num) => () => num);
        expect(
          printNodeTest(zodToTs(schema, { ctx: { ...ctx, isResponse: true } })),
        ).toMatchSnapshot();
      });

      test("should handle an error within the transformation", () => {
        const schema = z
          .number()
          .transform(() => assert.fail("this should be handled"));
        expect(
          printNodeTest(zodToTs(schema, { ctx: { ...ctx, isResponse: true } })),
        ).toMatchSnapshot();
      });
    });
  });
});
