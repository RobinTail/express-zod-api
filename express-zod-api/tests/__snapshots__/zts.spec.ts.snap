// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`zod-to-ts > Example > should produce the expected results 1`] = `
"{
    string: string;
    number: number;
    arrayOfObjects: {
        string: string;
    }[];
    boolean: boolean;
    circular: SomeType;
    circular2: SomeType;
    union: {
        number: number;
    } | "hi";
    enum: "hi" | "bye";
    intersectionWithTransform: (number & bigint) & (number & string);
    date: any;
    undefined?: undefined;
    null: null;
    void: undefined;
    any: any;
    unknown: unknown;
    never: never;
    optionalString?: string | undefined;
    nullablePartialObject: {
        string?: string | undefined;
        number?: number | undefined;
        fixedArrayOfString?: string[] | undefined;
        object?: {
            string: string;
        } | undefined;
    } | null;
    tuple: [
        string,
        number,
        {
            string: string;
        }
    ];
    tupleRest: [
        string,
        number,
        ...boolean
    ];
    record: Record<string, {
        object: {
            arrayOfUnions: ([
                string,
                {
                    string: string;
                }
            ] | bigint)[];
        };
    }>;
    map: any;
    set: any;
    intersection: (string & number) | bigint;
    promise: any;
    optDefaultString?: string | undefined;
    refinedStringWithSomeBullshit: (string | number) & (bigint | null);
    lazy: SomeType;
    discUnion: {
        kind: "circle";
        radius: number;
    } | {
        kind: "square";
        x: number;
    } | {
        kind: "triangle";
        x: number;
        y: number;
    };
    branded: string;
    catch: number;
    pipeline: string;
    readonly: string;
}"
`;

exports[`zod-to-ts > Issue #2352: intersection of objects having same prop %# > should deduplicate the prop with a same name 1`] = `
"{
    query: string;
}"
`;

exports[`zod-to-ts > Issue #2352: intersection of objects having same prop %# > should deduplicate the prop with a same name 2`] = `
"{
    query: string;
}"
`;

exports[`zod-to-ts > Issue #2352: intersection of objects having same prop %# > should deduplicate the prop with a same name 3`] = `
"{
    query: string;
}"
`;

exports[`zod-to-ts > Issue #2352: intersection of objects having same prop %# > should deduplicate the prop with a same name 4`] = `
"{
    query: {};
}"
`;

exports[`zod-to-ts > Issue #2352: intersection of objects having same prop %# > should not flatten the result for objects with a conflicting prop 0 1`] = `
"{
    query: string;
} & {
    query: number;
}"
`;

exports[`zod-to-ts > Issue #2352: intersection of objects having same prop %# > should not flatten the result for objects with a conflicting prop 1 1`] = `
"{
    query: {
        sub: string;
    };
} & {
    query: {
        sub: number;
    };
}"
`;

exports[`zod-to-ts > Issue #2352: intersection of objects having same prop %# > should not flatten the result for objects with a conflicting prop 2 1`] = `
"{
    query: string;
} & {
    query?: string | undefined;
}"
`;

exports[`zod-to-ts > Issue #2352: intersection of objects having same prop %# > should not flatten the result for objects with a conflicting prop 3 1`] = `
"{
    query: string;
} & {
    query?: string | undefined;
}"
`;

exports[`zod-to-ts > PrimitiveSchema (isResponse=false) > outputs correct typescript 1`] = `
"{
    string: string;
    number: number;
    boolean: boolean;
    date: any;
    undefined?: undefined;
    null: null;
    void: undefined;
    any: any;
    unknown: unknown;
    never: never;
}"
`;

exports[`zod-to-ts > PrimitiveSchema (isResponse=true) > outputs correct typescript 1`] = `
"{
    string: string;
    number: number;
    boolean: boolean;
    date: unknown;
    undefined?: undefined;
    null: null;
    void: undefined;
    any: any;
    unknown: unknown;
    never: never;
}"
`;

exports[`zod-to-ts > enums > handles enum-like literals 0 1`] = `"0 | 1 | 2"`;

exports[`zod-to-ts > enums > handles enum-like literals 1 1`] = `""apple" | "banana" | "cantaloupe""`;

exports[`zod-to-ts > ez.buffer() > should be Buffer 1`] = `"Buffer"`;

exports[`zod-to-ts > ez.raw() > should depict the raw property 1`] = `"Buffer"`;

exports[`zod-to-ts > z.array() > outputs correct typescript 1`] = `
"{
    id: number;
    value: string;
}[]"
`;

exports[`zod-to-ts > z.discriminatedUnion() > outputs correct typescript 1`] = `
"{
    kind: "circle";
    radius: number;
} | {
    kind: "square";
    x: number;
} | {
    kind: "triangle";
    x: number;
    y: number;
}"
`;

exports[`zod-to-ts > z.literal() > Should produce the correct typescript 0 1`] = `""test""`;

exports[`zod-to-ts > z.literal() > Should produce the correct typescript 1 1`] = `"true"`;

exports[`zod-to-ts > z.literal() > Should produce the correct typescript 2 1`] = `"false"`;

exports[`zod-to-ts > z.literal() > Should produce the correct typescript 3 1`] = `"123"`;

exports[`zod-to-ts > z.literal() > Should produce the correct typescript 4 1`] = `"undefined"`;

exports[`zod-to-ts > z.nullable() > outputs correct typescript 1`] = `
"{
    username: string | null;
}"
`;

exports[`zod-to-ts > z.object() > does not unnecessary quote identifiers 1`] = `
"{
    id: string;
    name: string;
    countryOfOrigin: string;
}"
`;

exports[`zod-to-ts > z.object() > escapes correctly 1`] = `
"{
    "\\\\": string;
    "\\"": string;
    "'": string;
    "\`": string;
    "\\n": number;
    $e: any;
    "4t": any;
    _r: any;
    "-r"?: undefined;
}"
`;

exports[`zod-to-ts > z.object() > specially handles coercive schema in response 1`] = `
"{
    prop: string;
}"
`;

exports[`zod-to-ts > z.object() > supports string literal properties 1`] = `
"{
    "5": number;
    "string-literal": string;
}"
`;

exports[`zod-to-ts > z.object() > supports zod.deprecated() 1`] = `
"{
    /** @deprecated */
    one: string;
    /** @deprecated with description */
    two: string;
}"
`;

exports[`zod-to-ts > z.object() > supports zod.describe() 1`] = `
"{
    /** The name of the item */
    name: string;
    /** The price of the item */
    price: number;
}"
`;

exports[`zod-to-ts > z.optional() > Zod 4: should add question mark only to optional props 1`] = `
"{
    optional?: string | undefined;
    required: string;
    transform?: number | undefined;
    or?: (number | string) | undefined;
    tuple?: [
        string,
        number,
        {
            optional?: string | undefined;
            required: string;
        }
    ] | undefined;
}"
`;

exports[`zod-to-ts > z.pipe() > transformations > should handle an error within the transformation 1`] = `"unknown"`;

exports[`zod-to-ts > z.pipe() > transformations > should handle preprocess error in request 1`] = `"any"`;

exports[`zod-to-ts > z.pipe() > transformations > should handle unsupported transformation in response 1`] = `"unknown"`;

exports[`zod-to-ts > z.pipe() > transformations > should produce the schema type 'intact' 1`] = `"number"`;

exports[`zod-to-ts > z.pipe() > transformations > should produce the schema type 'transformed' 1`] = `"string"`;

exports[`zod-to-ts > z.templateLiteral() > should produce the correct typescript 0 1`] = `"\`start\${number}mid\${boolean}end\`"`;

exports[`zod-to-ts > z.templateLiteral() > should produce the correct typescript 1 1`] = `"\`\${number}one\${boolean}two\`"`;

exports[`zod-to-ts > z.templateLiteral() > should produce the correct typescript 2 1`] = `""one-two-three""`;

exports[`zod-to-ts > z.templateLiteral() > should produce the correct typescript 3 1`] = `"\`\${string}\${number}\${boolean}\`"`;

exports[`zod-to-ts > z.templateLiteral() > should produce the correct typescript 4 1`] = `"\`\${number}\`"`;

exports[`zod-to-ts > z.templateLiteral() > should produce the correct typescript 5 1`] = `""only""`;

exports[`zod-to-ts > z.templateLiteral() > should produce the correct typescript 6 1`] = `"\`\${number}more\`"`;

exports[`zod-to-ts > z.templateLiteral() > should produce the correct typescript 7 1`] = `"\`leading\${boolean}\`"`;

exports[`zod-to-ts > z.templateLiteral() > should produce the correct typescript 8 1`] = `""""`;

exports[`zod-to-ts > z.templateLiteral() > should produce the correct typescript 9 1`] = `""head""`;

exports[`zod-to-ts > z.templateLiteral() > should produce the correct typescript 10 1`] = `""head0""`;
